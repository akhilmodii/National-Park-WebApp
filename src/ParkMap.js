import React, { Component } from 'react';
import { Map, Marker, Tooltip, TileLayer } from 'react-leaflet';
import L from 'leaflet';
import Park from './Park.js';
import { connect } from 'react-redux';
import { setParkCardToShow } from './actions';

class ParkMap extends Component {
  constructor(props) {
    super(props);
    this.state = {
      uniqueStateZoom: { Alaska: 4, California: 5, Michigan: 5 },
    };
  }

  getPark = (event) => {
    const { parks, setParkCardToShow } = this.props;
    const park = parks.find(park => park.parkCode === event.target.options.id);
    console.log(park)
    setParkCardToShow(park.parkCode);
  }

  findParksToShow = () => {
    const { parks, parksToDisplay, visitedParkCodes, bucketListParkCodes, usStates } = this.props;
    let parksToShow;
    switch (parksToDisplay) {
      case 'all':
        parksToShow = parks;
        break;
      case 'visited':
        parksToShow = parks.filter(park => visitedParkCodes.includes(park.parkCode));
        break;
      case 'bucket':
        parksToShow = parks.filter(park => bucketListParkCodes.includes(park.parkCode));
        break;
      default: 
        const state = usStates.find(state => state.name === parksToDisplay);
        parksToShow = this.showParksInState(state.abbreviation);
    }
    return this.showParks(parksToShow);
  }

  showParksInState = (stateCode) => {
    return this.props.parks.filter(park => park.states.includes(stateCode))
  }

  showParks = (parks) => {
    const icon = {
      iconSize: [25, 41],
      iconAnchor: [12, 41],
      popupAnchor: [1, -34],
      shadowSize: [41, 41],
      shadowUrl: './assets/marker-shadow.png'
    };
    const greenIcon = new L.icon({ ...icon, iconUrl: './assets/green-icon.png'});
    const violetIcon = new L.icon({ ...icon, iconUrl: './assets/purple-icon.png'});
    const blueIcon = new L.icon({ ...icon, iconUrl: './assets/red-icon.png'});

    const { visitedParkCodes, bucketListParkCodes } = this.props;
    return parks.map(park => {
      const { parkCode, name, lat, lon } = park;
      let currentIcon;
      if (visitedParkCodes.includes(parkCode)) {
        currentIcon = greenIcon;
      } else if (bucketListParkCodes.includes(parkCode)) {
        currentIcon = violetIcon;
      } else {
        currentIcon = blueIcon;
      }
      return (
        <Marker 
          position={[lat, lon]}
          icon={currentIcon}
          onClick={this.getPark}
          key={parkCode}
          id={parkCode}>
          <Tooltip>{name}</Tooltip>
        </Marker>
      )
    })
  }

  getMapInfo = () => {
    let zoom;
    let position;
    if (this.props.showUsState) {
      const { usStates, parksToDisplay } = this.props;
      const { uniqueStateZoom } = this.state;
      const state = usStates.find(state => state.name === parksToDisplay);
      zoom = uniqueStateZoom[state.name] ? uniqueStateZoom[state.name] : 6;
      position = [state.latitude, state.longitude];
    } else {
      zoom = 4;
      position = [37.0902, -95.7129];
    }
    return [zoom, position]
  }
  
  render() {
    const mapInfo = this.getMapInfo();
    const parksToShow = this.findParksToShow();
    const { currentParkCode, parks } = this.props;
    return (
     <div className="map-container">
        <Map 
          id="map" 
          minZoom='3'
          maxZoom='8'
          center={mapInfo[1]}
          zoom={mapInfo[0]}>
          <TileLayer 
            url='https://server.arcgisonline.com/ArcGIS/rest/services/World_Physical_Map/MapServer/tile/{z}/{y}/{x}'
            attribution='Tiles &copy; Esri &mdash; Source: US National Park Service'
            maxZoom='8'
          />
          <TileLayer 
            url='https://stamen-tiles-{s}.a.ssl.fastly.net/toner-hybrid/{z}/{x}/{y}{r}.{ext}'
            attribution='Map tiles by <a href="http://stamen.com">Stamen Design</a>, <a href="http://creativecommons.org/licenses/by/3.0">CC BY 3.0</a> &mdash; Map data &copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
            subdomains='abcd'
            minZoom='0'
            maxZoom='20'
            ext='png'
          />
          {parksToShow}
        </Map>
        {
          currentParkCode.length && 
          <Park 
            park={
              parks.find(park => park.parkCode === currentParkCode)
            }
          />
        }
      </div>   
    )
  }
}

const mapStateToProps = (state) => ({
  usStates: state.usStates,
  parksToDisplay: state.parksToDisplay,
  parks: state.parks,
  currentParkCode: state.currentParkCode,
  visitedParkCodes: state.visitedParkCodes,
  bucketListParkCodes: state.bucketListParkCodes,
  showUsState: state.showUsState,
});

const mapDispatchToProps = (dispatch) => ({
  setParkCardToShow: (parkCode) => dispatch(setParkCardToShow(parkCode)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ParkMap);
