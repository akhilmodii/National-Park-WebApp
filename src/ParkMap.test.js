import React from 'react';
import ParkMap from './ParkMap';
import { shallow } from 'enzyme';

const parks = [
  {
    parkName: "Death Valley",
    state: [
    "California",
    "Nevada"
    ],
    dateEstablished: "October 31, 1994",
    latitude: "36.24°N",
    longitude: "116.82°W",
    annualVisitors: 1294827,
    petsWelcome: true,
    editorsChoice: "Badwater Basin Salt Flats",
    websiteUrl: "https://www.nps.gov/deva/index.htm",
    urlCode: "deva",
    image: "assets/death-valley.jpg",
    sizeAcres: 337306314,
    description: "The Badlands are a collection of buttes, pinnacles, spires, and mixed-grass prairies. The White River Badlands contain the largest assemblage of known late Eocene and Oligocene mammal fossils. The wildlife includes bison, bighorn sheep, black-footed ferrets, and prairie dogs."
  },
  {
    parkName: "Great Smoky Mountains",
    state: [
    "Tennessee",
    "North Carolina"
    ],
    dateEstablished: "June 15, 1934",
    latitude: "35.68°N",
    longitude: "83.53°W",
    annualVisitors: 11338893,
    petsWelcome: true,
    editorsChoice: "Appalachian Trail",
    websiteUrl: "https://www.nps.gov/grsm/index.htm",
    urlCode: "grsm",
    image: "assets/great-smoky-mtns.jpeg",
    sizeAcres: 52242688,
    description: "The Great Smoky Mountains, part of the Appalachian Mountains, span a wide range of elevations, making them home to over 400 vertebrate species, 100 tree species, and 5000 plant species. Hiking is the park's main attraction, with over 800 miles (1,300 km) of trails, including 70 miles (110 km) of the Appalachian Trail. Other activities include fishing, horseback riding, and touring nearly 80 historic structures."
  },
  {
    parkName: "Yellowstone",
    state: [
    "Wyoming",
    "Montana",
    "Idaho"
    ],
    dateEstablished: "March 1, 1872",
    latitude: "44.60°N",
    longitude: "110.50°W",
    annualVisitors: 4116524,
    petsWelcome: true,
    editorsChoice: "Uncle Tom's Trail",
    websiteUrl: "https://www.nps.gov/yell/index.htm",
    urlCode: "yell",
    image: "assets/yellowstone.jpg",
    sizeAcres: 221979071,
    description: "Situated on the Yellowstone Caldera, the park has an expansive network of geothermal areas including boiling mud pots, vividly colored hot springs such as Grand Prismatic Spring, and regularly erupting geysers, the best-known being Old Faithful. The yellow-hued Grand Canyon of the Yellowstone River contains several high waterfalls, and four mountain ranges traverse the park. More than 60 mammal species including gray wolves, grizzly bears, black bears, lynxes, bison, and elk, make this park one of the best wildlife viewing spots in the country."
  }
]
const visitedParks = ['deva', 'grsm'];
const bucketListParks = ['yell'];
const updateParkCodesMock = jest.fn();
const stateName = 'default';
const stateCoord = [];


describe('ParkMap', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <ParkMap
        parks={parks}
        stateName={stateName}
        stateCoord={stateCoord}
        visitedParks={visitedParks}
        bucketListParks={bucketListParks}
        updateParkCodes={updateParkCodesMock}
      />
    );
  });
  
  it('should match the snapshot with all data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have the proper default state', () => {
    expect(wrapper.state()).toEqual({ selectedPark: null, zoom: 4, uniqueStateZoom: { Alaska: 4, California: 5, Michigan: 5 }, position: [37.0902, - 95.7129] } );
  });

  it('should update state when getPark is called', () => {
    expect(wrapper.state('selectedPark')).toEqual(null);
    wrapper.find('#deva').simulate('click', { target: { options: {id: "deva"} }});
    expect(wrapper.state('selectedPark')).toEqual(parks[0]);
  });
});