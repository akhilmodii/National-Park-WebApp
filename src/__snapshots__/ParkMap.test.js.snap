// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParkMap should match the snapshot with all data passed in 1`] = `
<div
  className="map-container"
>
  <Map
    center={
      Array [
        37.0902,
        -95.7129,
      ]
    }
    id="map"
    maxZoom="8"
    minZoom="3"
    zoom={4}
  >
    <ForwardRef(Leaflet(TileLayer))
      attribution="Tiles © Esri — Source: US National Park Service"
      maxZoom="8"
      url="https://server.arcgisonline.com/ArcGIS/rest/services/World_Physical_Map/MapServer/tile/{z}/{y}/{x}"
    />
    <ForwardRef(Leaflet(TileLayer))
      attribution="Map tiles by <a href=\\"http://stamen.com\\">Stamen Design</a>, <a href=\\"http://creativecommons.org/licenses/by/3.0\\">CC BY 3.0</a> — Map data © <a href=\\"http://www.openstreetmap.org/copyright\\">OpenStreetMap</a>"
      ext="png"
      maxZoom="20"
      minZoom="0"
      subdomains="abcd"
      url="https://stamen-tiles-{s}.a.ssl.fastly.net/toner-hybrid/{z}/{x}/{y}{r}.{ext}"
    />
    <ForwardRef(Leaflet(Marker))
      icon={
        NewClass {
          "_initHooksCalled": true,
          "options": Object {
            "iconAnchor": Array [
              12,
              41,
            ],
            "iconSize": Array [
              25,
              41,
            ],
            "iconUrl": "./assets/marker-icon-green.png",
            "popupAnchor": Array [
              1,
              -34,
            ],
            "shadowSize": Array [
              41,
              41,
            ],
            "shadowUrl": "./assets/marker-shadow.png",
          },
        }
      }
      id="deva"
      key="deva"
      onClick={[Function]}
      position={
        Array [
          "36.24",
          "-116.82",
        ]
      }
    >
      <ForwardRef(Leaflet(Tooltip))>
        Death Valley
      </ForwardRef(Leaflet(Tooltip))>
    </ForwardRef(Leaflet(Marker))>
    <ForwardRef(Leaflet(Marker))
      icon={
        NewClass {
          "_initHooksCalled": true,
          "options": Object {
            "iconAnchor": Array [
              12,
              41,
            ],
            "iconSize": Array [
              25,
              41,
            ],
            "iconUrl": "./assets/marker-icon-green.png",
            "popupAnchor": Array [
              1,
              -34,
            ],
            "shadowSize": Array [
              41,
              41,
            ],
            "shadowUrl": "./assets/marker-shadow.png",
          },
        }
      }
      id="grsm"
      key="grsm"
      onClick={[Function]}
      position={
        Array [
          "35.68",
          "-83.53",
        ]
      }
    >
      <ForwardRef(Leaflet(Tooltip))>
        Great Smoky Mountains
      </ForwardRef(Leaflet(Tooltip))>
    </ForwardRef(Leaflet(Marker))>
    <ForwardRef(Leaflet(Marker))
      icon={
        NewClass {
          "_initHooksCalled": true,
          "options": Object {
            "iconAnchor": Array [
              12,
              41,
            ],
            "iconSize": Array [
              25,
              41,
            ],
            "iconUrl": "./assets/marker-icon-violet.png",
            "popupAnchor": Array [
              1,
              -34,
            ],
            "shadowSize": Array [
              41,
              41,
            ],
            "shadowUrl": "./assets/marker-shadow.png",
          },
        }
      }
      id="yell"
      key="yell"
      onClick={[Function]}
      position={
        Array [
          "44.60",
          "-110.50",
        ]
      }
    >
      <ForwardRef(Leaflet(Tooltip))>
        Yellowstone
      </ForwardRef(Leaflet(Tooltip))>
    </ForwardRef(Leaflet(Marker))>
  </Map>
</div>
`;
